pkg/tsdb/loki/streaming.go

```diff
@@ -0,0 +1,201 @@
+		return &backend.SubscribeStreamResponse{
+			Status:      backend.SubscribeStreamStatusOK,
+			InitialData: msg,
+		}, err
+	}
+
+	// nothing yet
+	return &backend.SubscribeStreamResponse{
+		Status: backend.SubscribeStreamStatusOK,
+	}, err
+}
+
+// Single instance for each channel (results are shared with all listeners)
+func (s *Service) RunStream(ctx context.Context, req *backend.RunStreamRequest, sender *backend.StreamSender) error {
+	dsInfo, err := s.getDSInfo(req.PluginContext)
+	if err != nil {
+		return err
+	}
+
+	query, err := parseQueryModel(req.Data)
+	if err != nil {
+		return err
+	}
+	if query.Expr == "" {
+		return fmt.Errorf("missing expr in cuannel")
+	}
+
+	count := int64(0)
+
+	interrupt := make(chan os.Signal, 1)
+	signal.Notify(interrupt, os.Interrupt)
+
+	params := url.Values{}
+	params.Add("query", query.Expr)
+
+	isV1 := false
+	wsurl, _ := url.Parse(dsInfo.URL)
+
+	// Check if the v2alpha endpoint exists
+	wsurl.Path = "/loki/api/v2alpha/tail"
+	if !is400(dsInfo.HTTPClient, wsurl) {
+		isV1 = true
+		wsurl.Path = "/loki/api/v1/tail"
+	}
+
+	if wsurl.Scheme == "https" {
+		wsurl.Scheme = "wss"
+	} else {
+		wsurl.Scheme = "ws"
+	}
+	wsurl.RawQuery = params.Encode()
+
+	s.plog.Info("connecting to websocket", "url", wsurl)
+	c, r, err := websocket.DefaultDialer.Dial(wsurl.String(), nil)
+	if err != nil {
+		s.plog.Error("error connecting to websocket", "err", err)
+		return fmt.Errorf("error connecting to websocket")
+	}
+
+	defer func() {
+		dsInfo.streamsMu.Lock()
+		delete(dsInfo.streams, req.Path)
+		dsInfo.streamsMu.Unlock()
+		if r != nil {
+			_ = r.Body.Close()
+		}
+		err = c.Close()
+		s.plog.Error("closing loki websocket", "err", err)
+	}()
+
+	prev := data.FrameJSONCache{}
+
+	// Read all messages
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		for {
+			_, message, err := c.ReadMessage()
+			if err != nil {
+				s.plog.Error("websocket read:", "err", err)
+				return
+			}
+
+			frame := &data.Frame{}
+			if isV1 {
+				// fmt.Printf("\n\n%s\n", string(message))
+				frame, err = lokiBytesToLabeledFrame(message)
+			} else {
+				err = json.Unmarshal(message, &frame)
+			}
+
+			if err == nil && frame != nil {
+				next, _ := data.FrameToJSONCache(frame)
+				if next.SameSchema(&prev) {
+					err = sender.SendBytes(next.Bytes(data.IncludeDataOnly))
+				} else {
+					err = sender.SendFrame(frame, data.IncludeAll)
+				}
+				prev = next
+
+				// Cache the initial data
+				dsInfo.streamsMu.Lock()
+				dsInfo.streams[req.Path] = prev
+				dsInfo.streamsMu.Unlock()
+			}
+
+			if err != nil {
+				s.plog.Error("websocket write:", "err", err, "raw", message)
+				return
+			}
+		}
+	}()
+
+	ticker := time.NewTicker(time.Second * 60) //.Step)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-done:
+			s.plog.Info("socket done")
+			return nil
+		case <-ctx.Done():
+			s.plog.Info("stop streaming (context canceled)")
+			return nil
```

Comment:
(i'm probably missing something here, but just to be sure).
when `ctx.Done()` happens, will the read-messages-loop above get stopped too somehow?